src/App.tsx

```diff
@@ -41,38 +42,35 @@ const App: React.FC = () => {
   };
 
   useEffect(() => {
-    setIsLoading(true);
-    auth.onAuthStateChanged((user) => {
-      if (user) {
-        setUser(user);
-
-        Promise.all([loadHistory(user.uid), loadUserProfile(user.uid)])
-          .catch(console.warn)
-          .finally(() => setIsLoading(false));
-      }
-    });
+    if (user) {
+      setIsLoading(true);
+      Promise.all([loadHistory(user.uid), loadUserProfile(user.uid)])
+        .catch(console.warn)
+        .finally(() => setIsLoading(false));
+    }
 
     if (!user) {
       setIsLoading(true);
       getRedirectResult(auth)
+        .then((cred) => {
+          if (cred) {
+            setUser(cred.user);
+          }
+        })
         .catch(console.warn)
         .finally(() => setIsLoading(false));
     }
-  }, [auth, user]);
+  }, [user]);
 
   if (isLoading) {
-    return (
-      <div className="centered">
-        <Spinner size="lg" color="var(--logo-color)" />
-      </div>
-    );
+    return <CenteredSpinner />;
   }
 
   return (
     <>
       <div className="centered">
-        {user && userProfile ? (
-          <Text>Hello {userProfile.first}.</Text>
+        {user ? (
+          <Text>Hello {userProfile?.first ?? user.displayName}.</Text>
```

Comment:
This makes it so if a userProfile isn't found it doesn't just show the login button again, and falls back to `user.displayName`